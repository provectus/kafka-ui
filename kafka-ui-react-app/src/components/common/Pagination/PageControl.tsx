import cx from 'classnames';
import React from 'react';
import { Link } from 'react-router-dom';

export interface PageControlProps {
  current: boolean;
  url: string;
  page: number;
}

const PageControl: React.FC<PageControlProps> = ({ current, url, page }) => {
  const classNames = cx('pagination-link', {
    'is-current': current,
  });

  return (
    <li>
      <Link className={classNames} to={url} aria-label={`Goto page ${page}`}>
        {page}
      </Link>
    </li>
  );
};

export default PageControl;
