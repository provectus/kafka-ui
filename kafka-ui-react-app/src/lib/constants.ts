import { SelectOption } from 'components/common/Select/Select';
import { ConfigurationParameters } from 'generated-sources';

declare global {
  interface Window {
    basePath: string;
  }
}

export const BASE_PARAMS: ConfigurationParameters = {
  basePath: window.basePath || '',
  credentials: 'include',
  headers: {
    'Content-Type': 'application/json',
  },
};

export const TOPIC_NAME_VALIDATION_PATTERN = /^[a-zA-Z0-9._-]+$/;
export const SCHEMA_NAME_VALIDATION_PATTERN = /^[.,A-Za-z0-9_/-]+$/;

export const TOPIC_CUSTOM_PARAMS_PREFIX = 'customParams';
export const TOPIC_CUSTOM_PARAMS: Record<string, string> = {
  'compression.type': 'producer',
  'leader.replication.throttled.replicas': '',
  'message.downconversion.enable': 'true',
  'segment.jitter.ms': '0',
  'flush.ms': '9223372036854775807',
  'follower.replication.throttled.replicas': '',
  'segment.bytes': '1073741824',
  'flush.messages': '9223372036854775807',
  'message.format.version': '2.3-IV1',
  'file.delete.delay.ms': '60000',
  'max.compaction.lag.ms': '9223372036854775807',
  'min.compaction.lag.ms': '0',
  'message.timestamp.type': 'CreateTime',
  preallocate: 'false',
  'min.cleanable.dirty.ratio': '0.5',
  'index.interval.bytes': '4096',
  'unclean.leader.election.enable': 'true',
  'retention.bytes': '-1',
  'delete.retention.ms': '86400000',
  'segment.ms': '604800000',
  'message.timestamp.difference.max.ms': '9223372036854775807',
  'segment.index.bytes': '10485760',
};

export const MILLISECONDS_IN_WEEK = 604_800_000;
export const MILLISECONDS_IN_DAY = 86_400_000;
export const MILLISECONDS_IN_SECOND = 1_000;

export const NOT_SET = -1;
export const BYTES_IN_GB = 1_073_741_824;
export const BUILD_VERSION_PATTERN = /v\d.\d.\d/;

export const PER_PAGE = 25;
export const MESSAGES_PER_PAGE = '100';

export const GIT_REPO_LINK = 'https://github.com/provectus/kafka-ui';
export const GIT_REPO_LATEST_RELEASE_LINK =
  'https://api.github.com/repos/provectus/kafka-ui/releases/latest';

export const LOCAL_STORAGE_KEY_PREFIX = 'kafka-ui';

export enum AsyncRequestStatus {
  initial = 'initial',
  pending = 'pending',
  fulfilled = 'fulfilled',
  rejected = 'rejected',
}

export const QUERY_REFETCH_OFF_OPTIONS = {
  refetchOnMount: false,
  refetchOnWindowFocus: false,
  refetchIntervalInBackground: false,
};

// Cluster Form Constants
export const AUTH_OPTIONS: SelectOption[] = [
  { value: 'SASL/JAAS', label: 'SASL/JAAS' },
  { value: 'SASL/GSSAPI', label: 'SASL/GSSAPI' },
  { value: 'SASL/OAUTHBEARER', label: 'SASL/OAUTHBEARER' },
  { value: 'SASL/PLAIN', label: 'SASL/PLAIN' },
  { value: 'SASL/SCRAM-256', label: 'SASL/SCRAM-256' },
  { value: 'SASL/SCRAM-512', label: 'SASL/SCRAM-512' },
  { value: 'Delegation tokens', label: 'Delegation tokens' },
  { value: 'SASL/LDAP', label: 'SASL/LDAP' },
  { value: 'SASL/AWS IAM', label: 'SASL/AWS IAM' },
  { value: 'mTLS', label: 'mTLS' },
];

export const SECURITY_PROTOCOL_OPTIONS: SelectOption[] = [
  { value: 'SASL_SSL', label: 'SASL_SSL' },
  { value: 'SASL_PLAINTEXT', label: 'SASL_PLAINTEXT' },
];
export const METRICS_OPTIONS: SelectOption[] = [
  { value: 'JMX', label: 'JMX' },
  { value: 'PROMETHEUS', label: 'PROMETHEUS' },
];
