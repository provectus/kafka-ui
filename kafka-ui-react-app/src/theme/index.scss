@import 'src/theme/bulma_overrides';

#root, body, html {
  width: 100%;
  position: relative;
  margin: 0;
  font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', 'Roboto', 'Oxygen',
    'Ubuntu', 'Cantarell', 'Fira Sans', 'Droid Sans', 'Helvetica Neue',
    sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  background: repeating-linear-gradient(
    145deg,
    rgba(0,0,0,.003),
    rgba(0,0,0,.005) 5px,
    rgba(0,0,0,0) 5px,
    rgba(0,0,0,0) 10px
  ),
  repeating-linear-gradient(
    -145deg,
    rgba(0,0,0,.003),
    rgba(0,0,0,.005) 5px,
    rgba(0,0,0,0) 5px,
    rgba(0,0,0,0) 10px
  );
  background-color: $light;

}

code {
  font-family: source-code-pro, Menlo, Monaco, Consolas, 'Courier New',
    monospace;
}

.cursor-pointer {
  cursor: pointer;
}
