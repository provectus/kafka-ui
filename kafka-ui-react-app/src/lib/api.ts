import {
  KsqlApi,
  TopicsApi,
  SchemasApi,
  BrokersApi,
  MessagesApi,
  ClustersApi,
  Configuration,
  KafkaConnectApi,
  ConsumerGroupsApi,
  AuthorizationApi,
  ApplicationConfigApi,
} from 'generated-sources';
import { BASE_PARAMS } from 'lib/constants';

const apiClientConf = new Configuration(BASE_PARAMS);

export const ksqlDbApiClient = new KsqlApi(apiClientConf);
export const topicsApiClient = new TopicsApi(apiClientConf);
export const brokersApiClient = new BrokersApi(apiClientConf);
export const schemasApiClient = new SchemasApi(apiClientConf);
export const messagesApiClient = new MessagesApi(apiClientConf);
export const clustersApiClient = new ClustersApi(apiClientConf);
export const kafkaConnectApiClient = new KafkaConnectApi(apiClientConf);
export const consumerGroupsApiClient = new ConsumerGroupsApi(apiClientConf);
export const authApiClient = new AuthorizationApi(apiClientConf);
export const appConfigApiClient = new ApplicationConfigApi(apiClientConf);
