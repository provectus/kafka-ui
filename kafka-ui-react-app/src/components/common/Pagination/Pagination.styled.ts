import styled from 'styled-components';
import { Link } from 'react-router-dom';
import theme from 'theme/theme';

export const Wrapper = styled.nav`
  display: flex;
  align-items: flex-end;
  padding: 25px 16px;
  gap: 15px;

  & > ul {
    display: flex;
    align-items: flex-end;

    & > li:not(:last-child) {
      margin-right: 12px;
    }
  }
`;

export const PaginationLink = styled(Link)<{ $isCurrent: boolean }>`
  display: flex;
  justify-content: center;
  align-items: center;

  height: 32px;
  width: 33px;

  border-radius: 4px;
  border: 1px solid
    ${({ $isCurrent }) =>
      $isCurrent
        ? theme.paginationStyles.currentPage
        : theme.paginationStyles.borderColor.normal};
  background-color: ${({ $isCurrent }) =>
    $isCurrent
      ? theme.paginationStyles.currentPage
      : theme.paginationStyles.backgroundColor};
  color: ${theme.paginationStyles.color.normal};

  &:hover {
    border: 1px solid
      ${({ $isCurrent }) =>
        $isCurrent
          ? theme.paginationStyles.currentPage
          : theme.paginationStyles.borderColor.hover};
    color: ${(props) => props.theme.paginationStyles.color.hover};
    cursor: ${({ $isCurrent }) => ($isCurrent ? 'default' : 'pointer')};
  }
`;

export const PaginationButton = styled(Link)`
  display: flex;
  align-items: center;
  padding: 6px 12px;
  height: 32px;
  border: 1px solid ${theme.paginationStyles.borderColor.normal};
  border-radius: 4px;
  color: ${theme.paginationStyles.color.normal};

  &:hover {
    border: 1px solid ${theme.paginationStyles.borderColor.hover};
    color: ${theme.paginationStyles.color.hover};
    cursor: pointer;
  }
  &:active {
    border: 1px solid ${theme.paginationStyles.borderColor.active};
    color: ${theme.paginationStyles.color.active};
  }
  &:disabled {
    border: 1px solid ${theme.paginationStyles.borderColor.disabled};
    color: ${theme.paginationStyles.color.disabled};
    cursor: not-allowed;
  }
`;

export const DisabledButton = styled.button`
  display: flex;
  align-items: center;
  padding: 6px 12px;
  height: 32px;
  border: 1px solid ${theme.paginationStyles.borderColor.disabled};
  background-color: ${theme.paginationStyles.backgroundColor};
  border-radius: 4px;
  font-size: 16px;
  color: ${theme.paginationStyles.color.disabled};
`;
