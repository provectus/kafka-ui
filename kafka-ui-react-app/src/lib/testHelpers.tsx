import React, { PropsWithChildren, ReactElement, useMemo } from 'react';
import {
  MemoryRouter,
  MemoryRouterProps,
  Route,
  Routes,
} from 'react-router-dom';
import fetchMock from 'fetch-mock';
import { Provider } from 'react-redux';
import { ThemeProvider } from 'styled-components';
import { theme } from 'theme/theme';
import {
  render,
  renderHook,
  RenderOptions,
  waitFor,
} from '@testing-library/react';
import { AnyAction, Store } from 'redux';
import { RootState } from 'redux/interfaces';
import { configureStore } from '@reduxjs/toolkit';
import rootReducer from 'redux/reducers';
import {
  QueryClient,
  QueryClientProvider,
  UseQueryResult,
} from '@tanstack/react-query';
import { ConfirmContextProvider } from 'components/contexts/ConfirmContext';
import ConfirmationModal from 'components/common/ConfirmationModal/ConfirmationModal';
import { GlobalSettingsContext } from 'components/contexts/GlobalSettingsContext';
import { UserInfoRolesAccessContext } from 'components/contexts/UserInfoRolesAccessContext';

import { RolesType, modifyRolesData } from './permissions';

interface CustomRenderOptions extends Omit<RenderOptions, 'wrapper'> {
  preloadedState?: Partial<RootState>;
  store?: Store<Partial<RootState>, AnyAction>;
  initialEntries?: MemoryRouterProps['initialEntries'];
  userInfo?: {
    roles?: RolesType;
    rbacFlag: boolean;
  };
}

interface WithRouteProps {
  children: React.ReactNode;
  path: string;
}

export const expectQueryWorks = async (
  mock: fetchMock.FetchMockStatic,
  result: { current: UseQueryResult<unknown, unknown> }
) => {
  await waitFor(() => expect(result.current.isFetched).toBeTruthy());
  expect(mock.calls()).toHaveLength(1);
  expect(result.current.data).toBeDefined();
};

export const WithRoute: React.FC<WithRouteProps> = ({ children, path }) => {
  return (
    <Routes>
      <Route path={path} element={children} />
    </Routes>
  );
};

export const TestQueryClientProvider: React.FC<PropsWithChildren<unknown>> = ({
  children,
}) => {
  // use new QueryClient instance for each test run to avoid issues with cache
  const queryClient = new QueryClient({
    defaultOptions: { queries: { retry: false } },
  });
  return (
    <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>
  );
};

/**
 * @description it will create a UserInfo Provider that will actually
 * disable the rbacFlag , to user if you can pass it as an argument
 * */
const TestUserInfoProvider: React.FC<
  PropsWithChildren<{ data?: { roles?: RolesType; rbacFlag: boolean } }>
> = ({ children, data }) => {
  const contextValue = useMemo(() => {
    const roles = modifyRolesData(data?.roles);

    return {
      username: 'test',
      rbacFlag: !!(typeof data?.rbacFlag === 'undefined'
        ? false
        : data?.rbacFlag),
      roles,
    };
  }, [data]);

  return (
    <UserInfoRolesAccessContext.Provider value={contextValue}>
      {children}
    </UserInfoRolesAccessContext.Provider>
  );
};

const customRender = (
  ui: ReactElement,
  {
    preloadedState,
    store = configureStore<RootState>({
      reducer: rootReducer,
      preloadedState,
    }),
    initialEntries,
    userInfo,
    ...renderOptions
  }: CustomRenderOptions = {}
) => {
  // overrides @testing-library/react render.
  const AllTheProviders: React.FC<PropsWithChildren<unknown>> = ({
    children,
  }) => (
    <TestQueryClientProvider>
      <GlobalSettingsContext.Provider value={{ hasDynamicConfig: false }}>
        <ThemeProvider theme={theme}>
          <TestUserInfoProvider data={userInfo}>
            <ConfirmContextProvider>
              <Provider store={store}>
                <MemoryRouter initialEntries={initialEntries}>
                  <div>
                    {children}
                    <ConfirmationModal />
                  </div>
                </MemoryRouter>
              </Provider>
            </ConfirmContextProvider>
          </TestUserInfoProvider>
        </ThemeProvider>
      </GlobalSettingsContext.Provider>
    </TestQueryClientProvider>
  );
  return render(ui, { wrapper: AllTheProviders, ...renderOptions });
};

const customRenderHook = (hook: () => UseQueryResult<unknown, unknown>) =>
  renderHook(hook, { wrapper: TestQueryClientProvider });

export { customRender as render, customRenderHook as renderQueryHook };

export class EventSourceMock {
  url: string;

  close: () => void;

  open: () => void;

  error: () => void;

  onmessage: () => void;

  constructor(url: string) {
    this.url = url;
    this.open = jest.fn();
    this.error = jest.fn();
    this.onmessage = jest.fn();
    this.close = jest.fn();
  }
}
