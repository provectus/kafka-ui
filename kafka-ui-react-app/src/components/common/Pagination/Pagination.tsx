import { PER_PAGE } from 'lib/constants';
import usePagination from 'lib/hooks/usePagination';
import { range } from 'lodash';
import React from 'react';
import PageControl from 'components/common/Pagination/PageControl';

import * as S from './Pagination.styled';

export interface PaginationProps {
  totalPages: number;
}

const NEIGHBOURS = 2;

const Pagination: React.FC<PaginationProps> = ({ totalPages }) => {
  const { page, perPage, pathname } = usePagination();

  const currentPage = page || 1;
  const currentPerPage = perPage || PER_PAGE;

  const getPath = (newPage: number) =>
    `${pathname}?page=${Math.max(newPage, 1)}&perPage=${currentPerPage}`;

  const pages = React.useMemo(() => {
    // Total visible numbers: neighbours, current, first & last
    const totalNumbers = NEIGHBOURS * 2 + 3;
    // totalNumbers + `...`*2
    const totalBlocks = totalNumbers + 2;

    if (totalPages <= totalBlocks) {
      return range(1, totalPages + 1);
    }

    const startPage = Math.max(
      2,
      Math.min(currentPage - NEIGHBOURS, totalPages)
    );
    const endPage = Math.min(
      totalPages - 1,
      Math.min(currentPage + NEIGHBOURS, totalPages)
    );

    let p = range(startPage, endPage + 1);

    const hasLeftSpill = startPage > 2;
    const hasRightSpill = totalPages - endPage > 1;
    const spillOffset = totalNumbers - (p.length + 1);

    switch (true) {
      case hasLeftSpill && !hasRightSpill: {
        p = [...range(startPage - spillOffset - 1, startPage - 1), ...p];
        break;
      }

      case !hasLeftSpill && hasRightSpill: {
        p = [...p, ...range(endPage + 1, endPage + spillOffset + 1)];
        break;
      }

      default:
        break;
    }

    return p;
  }, []);

  return (
    <S.Wrapper role="navigation" aria-label="pagination">
      {currentPage > 1 ? (
        <S.PaginationButton to={getPath(currentPage - 1)}>
          Previous
        </S.PaginationButton>
      ) : (
        <S.DisabledButton disabled>Previous</S.DisabledButton>
      )}
      {totalPages > 1 && (
        <ul>
          {!pages.includes(1) && (
            <PageControl
              page={1}
              current={currentPage === 1}
              url={getPath(1)}
            />
          )}
          {!pages.includes(2) && (
            <li>
              <span className="pagination-ellipsis">&hellip;</span>
            </li>
          )}
          {pages.map((p) => (
            <PageControl
              key={`page-${p}`}
              page={p}
              current={p === currentPage}
              url={getPath(p)}
            />
          ))}
          {!pages.includes(totalPages - 1) && (
            <li>
              <span className="pagination-ellipsis">&hellip;</span>
            </li>
          )}
          {!pages.includes(totalPages) && (
            <PageControl
              page={totalPages}
              current={currentPage === totalPages}
              url={getPath(totalPages)}
            />
          )}
        </ul>
      )}
      {currentPage < totalPages ? (
        <S.PaginationButton to={getPath(currentPage + 1)}>
          Next
        </S.PaginationButton>
      ) : (
        <S.DisabledButton disabled>Next</S.DisabledButton>
      )}
    </S.Wrapper>
  );
};

export default Pagination;
