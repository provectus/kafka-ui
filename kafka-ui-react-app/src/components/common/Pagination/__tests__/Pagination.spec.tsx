import React from 'react';
import { StaticRouter } from 'react-router-dom';
import Pagination, {
  PaginationProps,
} from 'components/common/Pagination/Pagination';
import theme from 'theme/theme';
import { render } from 'lib/testHelpers';
import { screen } from '@testing-library/react';

describe('Pagination', () => {
  const setupComponent = (search = '', props: Partial<PaginationProps> = {}) =>
    render(
      <StaticRouter location={{ pathname: '/my/test/path/23', search }}>
        <Pagination totalPages={11} {...props} />
      </StaticRouter>
    );

  describe('next & prev buttons', () => {
    it('renders disable prev button and enabled next link', () => {
      setupComponent('?page=1');
      expect(screen.getByText('Previous')).toBeDisabled();
      expect(screen.getByText('Next')).toBeInTheDocument();
    });

    it('renders disable next button and enabled prev link', () => {
      setupComponent('?page=11');
      expect(screen.getByText('Previous')).toBeInTheDocument();
      expect(screen.getByText('Next')).toBeDisabled();
    });

    it('renders next & prev links with correct path', () => {
      setupComponent('?page=5&perPage=20');
      expect(screen.getByText('Previous')).toBeInTheDocument();
      expect(screen.getByText('Next')).toBeInTheDocument();
      expect(screen.getByText('Previous')).toHaveAttribute(
        'href',
        '/my/test/path/23?page=4&perPage=20'
      );
      expect(screen.getByText('Next')).toHaveAttribute(
        'href',
        '/my/test/path/23?page=6&perPage=20'
      );
    });
  });

  describe('spread', () => {
    it('renders 1 spread element after first page control', () => {
      setupComponent('?page=8');
      expect(screen.getAllByRole('listitem')[1]).toHaveTextContent('…');
      expect(screen.getAllByRole('listitem')[1].firstChild).toHaveClass(
        'pagination-ellipsis'
      );
    });

    it('renders 1 spread element before last spread control', () => {
      setupComponent('?page=2');
      expect(screen.getAllByRole('listitem')[7]).toHaveTextContent('…');
      expect(screen.getAllByRole('listitem')[7].firstChild).toHaveClass(
        'pagination-ellipsis'
      );
    });

    it('renders 2 spread elements', () => {
      setupComponent('?page=6');
      expect(screen.getAllByText('…').length).toEqual(2);
      expect(screen.getAllByRole('listitem')[0]).toHaveTextContent('1');
      expect(screen.getAllByRole('listitem')[1]).toHaveTextContent('…');
      expect(screen.getAllByRole('listitem')[7]).toHaveTextContent('…');
      expect(screen.getAllByRole('listitem')[8]).toHaveTextContent('11');
    });

    it('renders 0 spread elements', () => {
      setupComponent('?page=2', { totalPages: 8 });
      expect(screen.queryAllByText('…').length).toEqual(0);
      expect(screen.getAllByRole('listitem').length).toEqual(8);
    });
  });

  describe('current page', () => {
    it('check if it sets page 8 as current when page param is set', () => {
      setupComponent('?page=8');
      expect(screen.getByText('8')).toHaveStyle(
        `background-color: ${theme.pagination.currentPage}`
      );
    });

    it('check if it sets first page as current when page param not set', () => {
      setupComponent('', { totalPages: 8 });
      expect(screen.getByText('1')).toHaveStyle(
        `background-color: ${theme.pagination.currentPage}`
      );
    });
  });
});
