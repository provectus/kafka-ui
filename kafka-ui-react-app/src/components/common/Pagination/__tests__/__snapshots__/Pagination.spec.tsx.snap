// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination matches snapshot 1`] = `
<Pagination
  totalPages={11}
>
  <nav
    aria-label="pagination"
    className="pagination is-small is-right"
    role="navigation"
  >
    <button
      className="pagination-previous"
      disabled={true}
      type="button"
    >
      Previous
    </button>
    <Link
      className="pagination-next"
      to="/my/test/path/23?page=2&perPage=25"
    >
      <LinkAnchor
        className="pagination-next"
        href="/my/test/path/23?page=2&perPage=25"
        navigate={[Function]}
      >
        <a
          className="pagination-next"
          href="/my/test/path/23?page=2&perPage=25"
          onClick={[Function]}
        >
          Next page
        </a>
      </LinkAnchor>
    </Link>
    <ul
      className="pagination-list"
    >
      <PageControl
        current={true}
        page={1}
        url="/my/test/path/23?page=1&perPage=25"
      >
        <li>
          <Link
            aria-label="Goto page 1"
            className="pagination-link is-current"
            to="/my/test/path/23?page=1&perPage=25"
          >
            <LinkAnchor
              aria-label="Goto page 1"
              className="pagination-link is-current"
              href="/my/test/path/23?page=1&perPage=25"
              navigate={[Function]}
            >
              <a
                aria-label="Goto page 1"
                className="pagination-link is-current"
                href="/my/test/path/23?page=1&perPage=25"
                onClick={[Function]}
              >
                1
              </a>
            </LinkAnchor>
          </Link>
        </li>
      </PageControl>
      <PageControl
        current={false}
        key="page-2"
        page={2}
        url="/my/test/path/23?page=2&perPage=25"
      >
        <li>
          <Link
            aria-label="Goto page 2"
            className="pagination-link"
            to="/my/test/path/23?page=2&perPage=25"
          >
            <LinkAnchor
              aria-label="Goto page 2"
              className="pagination-link"
              href="/my/test/path/23?page=2&perPage=25"
              navigate={[Function]}
            >
              <a
                aria-label="Goto page 2"
                className="pagination-link"
                href="/my/test/path/23?page=2&perPage=25"
                onClick={[Function]}
              >
                2
              </a>
            </LinkAnchor>
          </Link>
        </li>
      </PageControl>
      <PageControl
        current={false}
        key="page-3"
        page={3}
        url="/my/test/path/23?page=3&perPage=25"
      >
        <li>
          <Link
            aria-label="Goto page 3"
            className="pagination-link"
            to="/my/test/path/23?page=3&perPage=25"
          >
            <LinkAnchor
              aria-label="Goto page 3"
              className="pagination-link"
              href="/my/test/path/23?page=3&perPage=25"
              navigate={[Function]}
            >
              <a
                aria-label="Goto page 3"
                className="pagination-link"
                href="/my/test/path/23?page=3&perPage=25"
                onClick={[Function]}
              >
                3
              </a>
            </LinkAnchor>
          </Link>
        </li>
      </PageControl>
      <PageControl
        current={false}
        key="page-4"
        page={4}
        url="/my/test/path/23?page=4&perPage=25"
      >
        <li>
          <Link
            aria-label="Goto page 4"
            className="pagination-link"
            to="/my/test/path/23?page=4&perPage=25"
          >
            <LinkAnchor
              aria-label="Goto page 4"
              className="pagination-link"
              href="/my/test/path/23?page=4&perPage=25"
              navigate={[Function]}
            >
              <a
                aria-label="Goto page 4"
                className="pagination-link"
                href="/my/test/path/23?page=4&perPage=25"
                onClick={[Function]}
              >
                4
              </a>
            </LinkAnchor>
          </Link>
        </li>
      </PageControl>
      <PageControl
        current={false}
        key="page-5"
        page={5}
        url="/my/test/path/23?page=5&perPage=25"
      >
        <li>
          <Link
            aria-label="Goto page 5"
            className="pagination-link"
            to="/my/test/path/23?page=5&perPage=25"
          >
            <LinkAnchor
              aria-label="Goto page 5"
              className="pagination-link"
              href="/my/test/path/23?page=5&perPage=25"
              navigate={[Function]}
            >
              <a
                aria-label="Goto page 5"
                className="pagination-link"
                href="/my/test/path/23?page=5&perPage=25"
                onClick={[Function]}
              >
                5
              </a>
            </LinkAnchor>
          </Link>
        </li>
      </PageControl>
      <PageControl
        current={false}
        key="page-6"
        page={6}
        url="/my/test/path/23?page=6&perPage=25"
      >
        <li>
          <Link
            aria-label="Goto page 6"
            className="pagination-link"
            to="/my/test/path/23?page=6&perPage=25"
          >
            <LinkAnchor
              aria-label="Goto page 6"
              className="pagination-link"
              href="/my/test/path/23?page=6&perPage=25"
              navigate={[Function]}
            >
              <a
                aria-label="Goto page 6"
                className="pagination-link"
                href="/my/test/path/23?page=6&perPage=25"
                onClick={[Function]}
              >
                6
              </a>
            </LinkAnchor>
          </Link>
        </li>
      </PageControl>
      <PageControl
        current={false}
        key="page-7"
        page={7}
        url="/my/test/path/23?page=7&perPage=25"
      >
        <li>
          <Link
            aria-label="Goto page 7"
            className="pagination-link"
            to="/my/test/path/23?page=7&perPage=25"
          >
            <LinkAnchor
              aria-label="Goto page 7"
              className="pagination-link"
              href="/my/test/path/23?page=7&perPage=25"
              navigate={[Function]}
            >
              <a
                aria-label="Goto page 7"
                className="pagination-link"
                href="/my/test/path/23?page=7&perPage=25"
                onClick={[Function]}
              >
                7
              </a>
            </LinkAnchor>
          </Link>
        </li>
      </PageControl>
      <li>
        <span
          className="pagination-ellipsis"
        >
          â€¦
        </span>
      </li>
      <PageControl
        current={false}
        page={11}
        url="/my/test/path/23?page=11&perPage=25"
      >
        <li>
          <Link
            aria-label="Goto page 11"
            className="pagination-link"
            to="/my/test/path/23?page=11&perPage=25"
          >
            <LinkAnchor
              aria-label="Goto page 11"
              className="pagination-link"
              href="/my/test/path/23?page=11&perPage=25"
              navigate={[Function]}
            >
              <a
                aria-label="Goto page 11"
                className="pagination-link"
                href="/my/test/path/23?page=11&perPage=25"
                onClick={[Function]}
              >
                11
              </a>
            </LinkAnchor>
          </Link>
        </li>
      </PageControl>
    </ul>
  </nav>
</Pagination>
`;
